File: src/main/java/com/gitlab/daring/image/command/transform/ThresholdCommands.java
package com.gitlab.daring.image.command.transform;

import com.gitlab.daring.image.command.Command;
import com.gitlab.daring.image.command.CommandRegistry;
import com.gitlab.daring.image.command.SimpleCommand;
import com.gitlab.daring.image.command.parameter.DoubleParam;
import com.gitlab.daring.image.command.parameter.EnumParam;
import com.gitlab.daring.image.command.parameter.IntParam;

import static com.gitlab.daring.image.util.ImageUtils.smat;
import static org.bytedeco.javacpp.opencv_core.inRange;
import static org.bytedeco.javacpp.opencv_imgproc.adaptiveThreshold;
import static org.bytedeco.javacpp.opencv_imgproc.threshold;

public class ThresholdCommands {

    public static void register(CommandRegistry r) {
        ThresholdCommands f = new ThresholdCommands();
        r.register("threshold", f::thresholdCommand);
        r.register("adaptiveThreshold", f::adaptiveThresholdCommand);
        r.register("inRange", f::inRangeCommand);
    }

    public Command thresholdCommand(String... ps) {
        SimpleCommand c = new SimpleCommand(ps);
        DoubleParam th = c.doubleParam(128, "0-255");
        DoubleParam mv = c.doubleParam(255, "0-255");
        EnumParam<ThresholdType> tp = c.enumParam(ThresholdType.class, ThresholdType.Bin);
        EnumParam<ThresholdFlag> fp = c.enumParam(ThresholdFlag.class, ThresholdFlag.None);
        return c.withFunc(m -> threshold(m, m, th.v, mv.v, tp.vi() + fp.vi() * 8));
    }

    public Command adaptiveThresholdCommand(String... ps) {
        SimpleCommand c = new SimpleCommand(ps);
        DoubleParam mv = c.doubleParam(255, "0-255");
        EnumParam<AdaptiveMethod> method = c.enumParam(AdaptiveMethod.class, AdaptiveMethod.Mean);
        EnumParam<ThresholdType> type = c.enumParam(ThresholdType.class, ThresholdType.Bin);
        IntParam bs = c.intParam(1, "0-50");
        IntParam cf = c.intParam(0, "0-100");
        return c.withFunc(m ->
            adaptiveThreshold(m, m, mv.v, method.vi(), type.vi(), bs.v * 2 + 1, cf.v)
        );
    }

    public Command inRangeCommand(String... ps) {
        SimpleCommand c = new SimpleCommand(ps);
        IntParam lb = c.intParam(0, "0-255");
        IntParam ub = c.intParam(255, "0-255");
        return c.withFunc(m -> inRange(m, smat(lb.v), smat(ub.v), m));
    }

    enum ThresholdType {Bin, BinInv, Trunc, ToZero, ToZeroInv}

    enum ThresholdFlag {None, OTSU, Triangle}

    enum AdaptiveMethod {Mean, Gaussian}

}

File: src/main/kotlin/com/gitlab/daring/image/command/transform/ThresholdCommands.kt
package com.gitlab.daring.image.command.transform

import com.gitlab.daring.image.command.Command
import com.gitlab.daring.image.command.CommandRegistry
import com.gitlab.daring.image.command.SimpleCommand
import com.gitlab.daring.image.util.ImageUtils.smat
import org.bytedeco.javacpp.opencv_core.inRange
import org.bytedeco.javacpp.opencv_imgproc.adaptiveThreshold
import org.bytedeco.javacpp.opencv_imgproc.threshold

internal object ThresholdCommands {

    @JvmStatic
    fun register(r: CommandRegistry) {
        r.register("threshold", this::thresholdCommand)
        r.register("adaptiveThreshold", this::adaptiveThresholdCommand)
        r.register("inRange", this::inRangeCommand)
    }

    fun thresholdCommand(args: Array<String>): Command {
        val c = SimpleCommand(args)
        val th = c.doubleParam(128.0, "0-255")
        val mv = c.doubleParam(255.0, "0-255")
        val tp = c.enumParam(ThresholdType.Bin)
        val fp = c.enumParam(ThresholdFlag.None)
        return c.withFunc { m -> threshold(m, m, th.v, mv.v, tp.vi + fp.vi * 8) }
    }

    fun adaptiveThresholdCommand(args: Array<String>): Command {
        val c = SimpleCommand(args)
        val mv = c.doubleParam(255.0, "0-255")
        val method = c.enumParam(AdaptiveMethod.Mean)
        val type = c.enumParam(ThresholdType.Bin)
        val bs = c.intParam(1, "0-50")
        val cf = c.intParam(0, "0-100")
        return c.withFunc { m -> adaptiveThreshold(m, m, mv.v, method.vi, type.vi, bs.v * 2 + 1, cf.dv) }
    }

    fun inRangeCommand(args: Array<String>): Command {
        val c = SimpleCommand(args)
        val lb = c.intParam(0, "0-255")
        val ub = c.intParam(255, "0-255")
        return c.withFunc { m -> inRange(m, smat(lb.v), smat(ub.v), m) }
    }

    enum class ThresholdType {
        Bin, BinInv, Trunc, ToZero, ToZeroInv
    }

    enum class ThresholdFlag {
        None, OTSU, Triangle
    }

    enum class AdaptiveMethod {
        Mean, Gaussian
    }

}
